src/gromacs/applied_forces/metatomic/metatomic_forceprovider.cpp

```diff
@@ -0,0 +1,564 @@
+    }
+    else
+    {
+        // non-main ranks do not load model or query capabilities, but they must still participate
+        // in collectives and must know dtype_ afterwards; dtype_ will be broadcast below.
+    }
+
+    // Initialize ML atom group data structures (must run on all ranks)
+    const auto& mtaIndices = options_.params_.mtaIndices_;
+    const int   n_atoms    = static_cast<int>(mtaIndices.size());
+
+    // Resize vectors that will be used on all ranks.
+    positions_.resize(n_atoms);
+    atomNumbers_.resize(n_atoms);
+    idxLookup_.resize(n_atoms); // Filled by gatherAtomNumbersIndices
+
+    // Atom numbers are static. We will populate them on the main rank
+    // and "broadcast" using sumReduce.
+
+    // Ensure non-main ranks have zeroed atomNumbers_ so a sumReduce acts like a broadcast.
+    std::fill(atomNumbers_.begin(), atomNumbers_.end(), 0);
+
+    // Now, only the main rank populates its version of the data.
+    if (mpiComm_.isMainRank())
+    {
+        for (int i = 0; i < n_atoms; ++i)
+        {
+            const int gIdx = mtaIndices[i];
+
+            if (gIdx >= options_.params_.numAtoms_)
+            {
+                GMX_THROW(APIError("Metatomic atom index " + std::to_string(gIdx) + " is out of bounds for topology numAtoms="
+                                   + std::to_string(options_.params_.numAtoms_)));
+            }
+            atomNumbers_[i] = options_.params_.atoms_.atom[gIdx].atomnumber;
+        }
+    }
+
+    // "Broadcast" the static atom numbers to all ranks: all ranks must call this.
+    if (mpiComm_.isParallel())
+    {
+        // resize/allocate on non-main ranks and broadcast raw bytes of vector
+        nblock_abc(mpiComm_.isMainRank(), mpiComm_.comm(), static_cast<std::size_t>(n_atoms), &atomNumbers_);
+    }
+
+    // Broadcast dtype info (so all ranks know whether the model uses float32 or float64).
+    // 0 = float64, 1 = float32
+    int dtype_code_local = 0;
+    if (mpiComm_.isMainRank())
+    {
+        dtype_code_local = (dtype_ == torch::kFloat32) ? 1 : 0;
+    }
+    // comm() may be MPI_COMM_NULL for SingleRank so..
+    if (mpiComm_.isParallel())
+    {
+        block_bc(mpiComm_.comm(), dtype_code_local);
+    }
+    // Set dtype_ on all ranks
+    dtype_ = (dtype_code_local == 1) ? torch::kFloat32 : torch::kFloat64;
+
+    // Initialize the lookup table. It will be populated correctly
+    // by the first call from the SimulationRunNotifier.
+    std::fill(idxLookup_.begin(), idxLookup_.end(), -1);
+
+    if (mpiComm_.isMainRank())
+    {
+        GMX_LOG(logger_.info)
+                .asParagraph()
+                .appendText("MetatomicForceProvider initialization complete.");
+    }
+}
+
+void MetatomicForceProvider::gatherAtomNumbersIndices()
+{
+    // This function is called on domain decomposition.
+    // Its ONLY job is to update the `idxLookup_` map, which maps
+    // the contiguous ML atom index (0..n_atoms-1) to the
```

Comment:
I don't understand what you mean here. If all ranks have access to all positions, then we can give all atoms to the model in each rank, and use `selected_atoms` to pass the sparse index/what the model in a given rank should be doing.

Reply:
Needs some speed testing, this PR is way too slow as written.